kart/help.py

```diff
@@ -0,0 +1,115 @@
+        )
+
+
+def get_renderer():
+    """
+    Return the appropriate HelpRenderer implementation for the
+    current platform.
+    """
+    if platform.system() == "Linux" or platform.system() == "Darwin":
+        return PosixHelpRenderer()
+
+
+class PagingHelpRenderer(object):
+    """
+    Interface for a help renderer.
+
+    The renderer is responsible for displaying the help content on
+    a particular platform.
+    """
+
+    def __init__(self, output_stream=sys.stdout):
+        self.output_stream = output_stream
+
+    PAGER = None
+
+    def get_pager_cmdline(self):
+        pager = self.PAGER
+        if "MANPAGER" in os.environ:
+            pager = os.environ["MANPAGER"]
+        elif "PAGER" in os.environ:
+            pager = os.environ["PAGER"]
+        return shlex.split(pager)
+
+    def render(self, ctx, contents):
+        """
+        Each implementation of HelpRenderer must implement this
+        render method.
+        """
+        # we add a dummy header now and remove it after conversion
+        header = "KART\n====\n"
+        converted_content = self._convert_doc_content(ctx, header + contents)
+        self._send_output_to_pager(converted_content)
+
+    def _send_output_to_pager(self, output):
+        cmdline = self.get_pager_cmdline()
+        L.debug("Running command: %s", cmdline)
+        p = self._popen(cmdline, stdin=PIPE)
+        p.communicate(input=output)
+
+    def _popen(self, *args, **kwargs):
+        return subprocess.Popen(*args, **kwargs)
+
+    def _convert_doc_content(self, ctx, contents):
+        return contents
+
+
+class PosixHelpRenderer(PagingHelpRenderer):
+    """
+    Render help content on a Posix-like system.  This includes
+    Linux and MacOS X.
+    """
+
+    PAGER = "less -R"
+
+    def _convert_doc_content(self, ctx, contents):
+        man_contents = publish_string(contents, writer=manpage.Writer()).decode(
+            "utf-8"
+        )[718:]
+        man_page = bytes(generate_man_page(ctx=ctx, other=man_contents), "utf-8")
+        if not self._exists_on_path("groff"):
+            raise ExecutableNotFoundError("groff")
+        cmdline = ["groff", "-m", "man", "-T", "ascii"]
+        L.debug("Running command: %s", cmdline)
+        p3 = self._popen(cmdline, stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        groff_output = p3.communicate(input=man_page)[0]
+        return groff_output
+
+    def _send_output_to_pager(self, output):
+        cmdline = self.get_pager_cmdline()
+        if not self._exists_on_path(cmdline[0]):
+            L.debug("Pager '%s' not found in PATH, printing raw help." % cmdline[0])
+            self.output_stream.write(output.decode("utf-8") + "\n")
+            self.output_stream.flush()
+            return
+        L.debug("Running command: %s", cmdline)
+        p = self._popen(cmdline, stdin=PIPE)
+        p.communicate(input=output)
+
+    def _exists_on_path(self, name):
```

Comment:
`shutil.which("groff")` seems to check the PATH of the Python runtime which may be different from the shell PATH.

Reply:
Have you seen this give different results to checking os.environ.get("PATH")? I didn't know that there could be more than one PATH variable in the environment at once. Since you're running "groff" directly - not via any shell - I would expect shutil.which to do a good job of finding the groff that will be run

Reply:
Ah, checked again with a fresh build, works perfectly. Don't know what went wrong before.